### PR TITLE
[JIT] Add early returns to JIT

### PR DESCRIPTION
**Pull request by eellison**
Add early returns to JIT with minimal changes to compiler.cpp and an IR->IR pass that will transform the graph so that there is only one return value. 

In compiler.cpp, record when a block will exit so that in the following example will work:
```
if cond:
    a = torch.zeros([2])
else: 
    return 2
a += 2
...
```
To match block outputs with values that will not be used, like in the above example with `a`, I add a Bottom Type that subtypes everything else. This allows shape propagation to continue to work, and makes it so that we don't need many extra nodes filling up the graph.

The IR transform currently doesn't work on Loops, I didn't add that to this PR to avoid too much complexity, but will add it as a stack (and it should be very little extra code). the IR  transform is commented at the top of the file.

### CODE REVIEW

**Comment by zdevito** [comment]
Haven't reviewed for correctness yet. But a high-level comment is that this is making substantial changes to the state and behavior of the compiler. I don't think I see support for loops yet, but already everywhere I see if statements in the compiler there are changes, and I see new function-level state added to the compiler pass. This won't scale as this expands to supporting more control flow.

**Comment by eellison** [comment]
@zdevito adding loops won’t require any more changes in the compiler, and should only be about 20 LOC in the pass. I will add it and re request review

**Comment by zdevito** [comment]
Still the compiler changes are already present are touching the environment/variable definition components of the compiler that are already too complicated.

**Comment by eellison** [comment]
Okay I will try to refactor a bit

**Comment by eellison** [comment]
@zdevito i refactored the compiler.cpp changes, and I think it is a lot easier to follow &  more maintainable now.

**Comment by zdevito** [comment]
Thanks! I haven't looked at everything yet (hopefully tonight, traveling today), but the change to compiler.cpp are much more understandable now.

**Comment by zdevito** [comment]
Still backed up, I will try to get this unblocked tomorrow.

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 1]
```diff
@@ -723,6 +724,10 @@ void AliasDb::analyzeBroadcastingChunk(Node* node) {
 
 // Register the fact that `value` is a pointer to `to`
 void AliasDb::makePointerTo(const Value* from, const Value* to) {
+  if (from->type() == BottomType::get() || to->type() == BottomType::get()) {
```
Comment (zdevito) [inline torch/csrc/jit/passes/alias_analysis.cpp]: Please add a comment.

Comment (zdevito) [reply torch/csrc/jit/passes/alias_analysis.cpp]: It is weird that this code is accumulating corner cases suggestions, @suo ?

File: torch/csrc/jit/passes/python_print.cpp [hunk 2]
```diff
@@ -489,12 +490,34 @@ struct PythonPrintPass {
     stmt << end;
   }
 
+  template <typename T>
```
Comment (zdevito) [inline torch/csrc/jit/passes/python_print.cpp]: What!? This will be really confusing when debugging if the order of loop carried dependencies now becomes a property of the debug names given to them. It is probably possible to eliminate the jitter that this is probably fixing in another way.

Comment (zdevito) [reply torch/csrc/jit/passes/python_print.cpp]: It is also very possible this is a bug. `useOf` for single-use things is not an identifier and it is possible they have side effects.

File: torch/csrc/jit/register_prim_ops.cpp [hunk 3]
```diff
@@ -867,6 +867,14 @@ RegisterOperators reg(
            }
            return 0;
          }),
+     Operator(
+         "prim::Bottom() -> Bottom",
+         [](Stack& stack) {
+           // Todo: add special unitialized ivalue that will error
+           // whenever it is used
+           push(stack, 1);
```
Comment (zdevito) [inline torch/csrc/jit/register_prim_ops.cpp]: Pushing 1 will lead to incredibly confusing errors. None would be a less surprising choice.

File: torch/csrc/jit/script/compiler.cpp [hunk 4]
```diff
@@ -569,15 +568,40 @@ struct to_ir {
...
   }
 
-  FunctionSchema emitDef(
-      const Def& def,
-      const Self& self,
-      Block* block) {
+  // If the graph might not return, at an implicit None return at the end
+  void handleMaybeNoReturn(const Def& def, Block* block) {
+    // if we're emitting a block that isn't the graph block,
+    // (like in a closure), do not add a return
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: What about a closure that doesn't end with a return and needs an implicit return?

File: torch/csrc/jit/script/compiler.cpp [hunk 5]
```diff
@@ -569,15 +568,40 @@ struct to_ir {
...
+    if (block != block->owningGraph()->block()) {
+      return;
+    }
+
+    if (exit_blocks.count(graph->block()) == 0) {
+      auto decl_ret = def_stack_.back().declared_return_type_;
+      if (decl_ret && decl_ret != NoneType::get() &&
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: If someone intents to return an Optional[T], I don't think they are ever going to simply omit the `return None`. I'd rather be more strict and just say that a function returning a value is missing a return statement.

File: torch/csrc/jit/script/compiler.cpp [hunk 6]
```diff
@@ -587,8 +611,9 @@ struct to_ir {
         emitFormalArguments(def, self, schema, block);
 
     // body
-    auto stmts_list = moveAllReturnsToEnd(def.statements());
+    auto stmts_list = def.statements();
     emitStatements(stmts_list.begin(), stmts_list.end());
+    handleMaybeNoReturn(def, block);
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: probably needs to happen for closures as well.

File: torch/csrc/jit/script/compiler.cpp [hunk 7]
```diff
@@ -1180,20 +1206,25 @@ struct to_ir {
...
     }
     for (auto& v : save_false->definedVariables()) {
-      if (save_true->findInAnyFrame(v)) {
+      if (save_true->findInAnyFrame(v) || true_exits) {
         mutated_variables.insert(v);
       }
     }
 
+    WithInsertPoint guard(graph->block()->nodes().front());
+    auto b_val = graph->insertNode(graph->create(prim::Bottom, {}, 1))
```
Comment (zdevito) [inline torch/csrc/jit/script/compiler.cpp]: prim::Constant?

Comment (zdevito) [reply torch/csrc/jit/script/compiler.cpp]: Why not just create one of these once at the beginning of the pass?

File: torch/csrc/jit/script/final_returns.cpp [hunk 8]
```diff
@@ -1,106 +1,288 @@
...
+// sentinel output is the output
+constexpr size_t RETURN_OFFSET = 2;
+constexpr size_t SENTINEL_OFFSET = 1;
 
-void checkNoReturn(const TreeRef& ref) {
-  if (ref->kind() == TK_RETURN)
-    throw ErrorReport(ref) << "return is not allowed from a loop.";
-  for (const TreeRef& child : ref->trees()) {
-    checkNoReturn(child);
+struct EarlyReturns {
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: Given what this pass will eventually become, EarlyReturns is not a great name.

File: torch/csrc/jit/script/final_returns.cpp [hunk 9]
```diff
@@ -1,106 +1,288 @@
...
+ * For blocks and control flow nodes that have a return statement that may have
+ * been hit, we add an extra output for the return value, and an extra output
+ * indicating whether or not the return has been hit (a sentinel value).
+ *
+ * When we encounter a node that might return, we guard all subsequent nodes
+ * in the block with the sentinel value of that node.
+ */
+
+// Will a block or node return
+enum Return_Status { WONT_RETURN, MIGHT_RETURN, WILL_RETURN };
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: nit: ReturnStatus

File: torch/csrc/jit/script/final_returns.cpp [hunk 10]
```diff
@@ -1,106 +1,288 @@
...
+
+// The return output of Control Flow Nodes is the second to last output,
+// sentinel output is the output
+constexpr size_t RETURN_OFFSET = 2;
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: This makes it seem like _every_ control node will have both these outputs. But only some will.

File: torch/csrc/jit/script/final_returns.cpp [hunk 11]
```diff
@@ -1,106 +1,288 @@
...
-          }
-          changed.emplace_back(
-              if_stmt.withNewBranches(true_final.stmts_, rest_final.stmts_));
-          return {true, List<Stmt>::unsafeCreate(range, std::move(changed))};
-        }
+  Value* getReturnVal(Block* b) {
+    return b->outputs().at(b->outputs().size() - RETURN_OFFSET);
+  }
+
+  Value* getSentinelVal(Node* n) {
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: Sentinel is not a great term because in the future there will be more than one sential (maybe one for the function having returned, and one for break/continue). `HasReturned` might be more specific.

File: torch/csrc/jit/script/final_returns.cpp [hunk 12]
```diff
@@ -1,106 +1,288 @@
...
+    AT_ASSERT(block_status[block] == WILL_RETURN);
+    block->registerOutput(block_return_vals[block]);
+  }
 
+  void run() {
+    makeReturnsFinal(graph->block());
+    setGraphOutput();
+  }
+
+  // After a call to makeReturnsFinal, a block will have set its return status
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: trailing underscores please -- it is hard to follow otherwise

File: torch/csrc/jit/script/final_returns.cpp [hunk 13]
```diff
@@ -1,106 +1,288 @@
...
+  std::unordered_map<Block*, Return_Status> block_status;
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: nit: one map containing a struct would be clearer

File: torch/csrc/jit/script/final_returns.cpp [hunk 14]
```diff
@@ -1,106 +1,288 @@
...
+          block_status[block] == WILL_RETURN &&
+          block->nodes().begin() == block->nodes().end());
     }
   }
-  if (return_none) {
-    // add an implicit return none node
-    changed.emplace_back(
-        Return::create(range, Expr(Compound::create(TK_NONE, range, {}))));
+
+  Value* getBottomVal() {
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: If the compiler universinoally created and passed this in, you could avoid having to repeat this whole thing here.

File: torch/csrc/jit/script/final_returns.cpp [hunk 15]
```diff
@@ -1,106 +1,288 @@
...
+    return b->outputs().at(b->outputs().size() - SENTINEL_OFFSET);
+  }
+
+  void registerReturnAndSentinel(Block* block, Value* ret, Value* sent) {
+    block->registerOutput(ret);
+    AT_ASSERT(sent->type() == BoolType::get());
+    block->registerOutput(sent);
+  }
+
+  void addReturnAndSentinel(Block* block) {
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: Soo many hash tables, Why not just one? The status itself could simply be figured out if the sentinel value.

File: torch/csrc/jit/script/final_returns.cpp [hunk 16]
```diff
@@ -1,106 +1,288 @@
...
+    if (b_status == WONT_RETURN) {
+      registerReturnAndSentinel(block, getBottomVal(), getBoolVal(false));
+    } else if (b_status == WILL_RETURN) {
+      registerReturnAndSentinel(
+          block, block_return_vals[block], getBoolVal(true));
+    } else if (b_status == MIGHT_RETURN) {
+      registerReturnAndSentinel(
+          block, block_return_vals[block], block_sentinel_val[block]);
+    } else {
+      AT_ASSERT(false);
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: nit: switch

File: torch/csrc/jit/script/final_returns.cpp [hunk 17]
```diff
@@ -1,106 +1,288 @@
...
+
+  // Recurses on the if node and returns its return status
+  // If status != WONT_RETURN, sets the block_return_val and sentinel val
+  // of its parent block before exit
+  Return_Status handleIf(Node* node) {
+    auto true_block = node->blocks().at(0);
+    auto false_block = node->blocks().at(1);
+
+    // recurse
+    makeReturnsFinal(true_block);
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: can't this just return the status?

File: torch/csrc/jit/script/final_returns.cpp [hunk 18]
```diff
@@ -1,106 +1,288 @@
...
+    return MIGHT_RETURN;
+  }
+
+  // Guards the remaining nodes in the block with an if node that takes
+  // sentinel as its conditional
+  Return_Status guardBlockNodes(
+      Block* block,
+      Value* sentinel,
+      Value* ret,
+      generic_graph_node_list_iterator<Node>& iter) {
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: We don't have a better type alias for this thing?

File: torch/csrc/jit/script/final_returns.cpp [hunk 19]
```diff
@@ -1,106 +1,288 @@
...
+    auto guard_block = new_if->addBlock();
+
+    // NB: need to set return_block status and return value before recursing
+    // or an empty block will appear to be a WONT_RETURN block
+    block_status[return_block] = WILL_RETURN;
+    block_return_vals[return_block] = ret;
+
+    // Move all remaining nodes into the guard block
+    WithInsertPoint b(guard_block);
+    auto insert_p = graph->insertConstant(false)->node();
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: wat?

Comment (zdevito) [reply torch/csrc/jit/script/final_returns.cpp]: You can just always say node->moveBefore(guard_block->return_node());

File: torch/csrc/jit/script/final_returns.cpp [hunk 20]
```diff
@@ -1,106 +1,288 @@
...
+          ->setUniqueName(block->outputs().at(i)->uniqueNameBase());
+    }
+    while (block->outputs().size() > 0) {
+      block->eraseOutput(0);
+    }
+    for (auto out : new_if->outputs()) {
+      block->registerOutput(out);
+    }
+
+    return handleIf(new_if);
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: General concern about stack overflow here. tail-call will probably happen, but otherwise the depth of this is the number of times someone writes a subblock with return. Probably not worth fixing yet.

File: torch/csrc/jit/script/final_returns.cpp [hunk 21]
```diff
@@ -1,106 +1,288 @@
...
+  void deleteAfterReturnNodes(Block* block, Node* return_node) {
+    auto nodes = block->nodes().reverse();
+    for (auto it = nodes.begin(); it != nodes.end() && *it != return_node;) {
+      auto node = it;
+      it++;
+      if (*node != block->return_node()) {
+        node->destroy();
+      }
+    }
+    if (return_node->kind() == prim::ReturnStmt) {
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: Feels weird that this cleanup is conditionally handled in here.

File: torch/csrc/jit/script/final_returns.cpp [hunk 22]
```diff
@@ -1,106 +1,288 @@
...
+    }
+  }
+
+  void makeReturnsFinal(Block* block) {
+    auto ret_status = WONT_RETURN;
+    for (auto it = block->nodes().begin(); it != block->nodes().end();) {
+      Node* node = *it;
+      it++;
+      switch (node->kind()) {
+        case prim::ReturnStmt: {
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: Makes more sense to delete the return statement in here.

File: torch/csrc/jit/script/final_returns.cpp [hunk 23]
```diff
@@ -1,106 +1,288 @@
...
   }
-  // we reach the end of the block, no returns have happened
-  // unless we just inserted a return_none implicit return.
-  return {return_none, List<Stmt>::unsafeCreate(range, std::move(changed))};
-}
 
-List<Stmt> moveAllReturnsToEnd(const List<Stmt>& stmts) {
-  return makeReturnsFinal(stmts, true).stmts_;
-}
+  Value* getBoolVal(bool val) {
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: Can't we just create them upfront universally and then let dce clean it up? We already run dce in lower tuples which will run right after this.

**Review by zdevito — COMMENTED** [review]
Cool! Overall approach looks good. A few notes to clarify the data structures in the pass. Also the jitter issue will need to be solved in a different way.

**Review by eellison — COMMENTED** [review]
Made most of the changes you suggested. Couple comments on the jitter / data structures

File: torch/csrc/jit/passes/python_print.cpp (hunk 2 above)
Comment (zdevito) [reply torch/csrc/jit/passes/python_print.cpp]: This is reordering the order in which useOf is printed. The text that useOf contains can have side effects in it.

Comment (zdevito) [reply torch/csrc/jit/passes/python_print.cpp]: Using debug names (which can be anything) is different from using variable names (which are not modified by any passes and are exactly as the user wrote). Change debug names should never change the order of things in the graph.

File: torch/csrc/jit/script/compiler.cpp (hunk 7 above)
Comment (zdevito) [reply torch/csrc/jit/script/compiler.cpp]: prim::Constant : All other constant values are generated by the prim::Constant node (e.g. there is no prim::None node to generate a None type). Same should happen for the Bottom type.

File: torch/csrc/jit/script/final_returns.cpp [hunk 24]
```diff
@@ -1,106 +1,270 @@
...
+              block, getHasReturned(node), getReturnVal(node), it);
+        }
         break;
+      }
+    }
+    if (block_status_.count(block) == 0) {
+      block_status_[block] = ret_status;
+      return ret_status;
+    } else {
+      // Guarded return blocks have their status set prior
```
Comment (zdevito) [inline torch/csrc/jit/script/final_returns.cpp]: Answer 'why' not 'what'. I can see that return status is already set, but I don't know why.

**Review by zdevito — COMMENTED** [review]
This looks good except for the jitter issues, which needs to be resolved in another way. I added more detailed comments there. I am not a fan of the data structures in the pass. Is there a timeline that you can commit to for fixing them if you want to do them in a future PR?

**Comment by eellison** [comment]
Solving jitter in PR https://github.com/pytorch/pytorch/pull/20015

**Comment by eellison** [comment]
Second to last commit removes has_returned struct. The last commit replaces prim::Bottom with prim::Uninitialized. I will need prim::Uninitialized to do loop returns so it makes since to do it here for bottom type as well.

You need it for loop returns because if you do something like:
```
for i in range(x):
   return  i
```

when you export the graph `i` needs to be set outside the for loop in order to be a loop carried dependency, and you can't use BottomType because the value set in the loop needs to subtype the existing value.

**Comment by zdevito** [comment]
prim::Uninitialized -- please don't feature creep on a diff. This is an 800 line change I would have to re-review to figure out how Uninitialied would fit in.

**Comment by eellison** [comment]
> prim::Uninitialized -- please don't feature creep on a diff. This is an 800 line change I would have to re-review to figure out how Uninitialied would fit in.

Understood feature creep on already long diff is frustrating. I didn't want to introduce a specific pattern that we couldn't change without BC breakage when prim::Uninitialized is the more general case. I can revert the commit if requested. 

As far as this diff goes, there shouldn't be any changes in behavior, since both nodes return a single BottomType value that errors on usage, and are not user exposed.

This will be helpful in returning from loops, as mentioned in comment ^^, because now this op can represent other values added by the compiler that error on usage.


-

**Comment by eellison** [comment]
Separately, I've talked with @suo about alias_analysis changes - he will review those.

File: torch/csrc/jit/passes/alias_analysis.cpp [hunk 25]
```diff
@@ -107,7 +109,9 @@ bool AliasDb::hasWriters(const Value* v) const {
     return numWrites_ != 0;
   }
 
-  if (!elementMap_.count(v) || v->mustBeNone()) {
+  // bc this is important api, check if value is non-aliasing before wildcard
+  // check
+  if (!elementMap_.count(v) || nonAliasingValue(v)) {
```
Comment (suo) [inline torch/csrc/jit/passes/alias_analysis.cpp]: do we still need this change? Below in `makePointerTo` you are a giving aliases to non-aliasing values anyway

File: torch/csrc/jit/passes/alias_analysis.h [hunk 26]
```diff
@@ -179,6 +181,9 @@ class AliasDb {
   // or does the element contain an element for which that's true
   bool cannotCheckAliasContainment(const Value* elem) const;
 
+  // Is this a value which will not alias
```
Comment (suo) [inline torch/csrc/jit/passes/alias_analysis.h]: Please add a more descriptive comment. There's a lot of subtlety because `nonAliasingValue` is not the same as `!shouldAnnotate()`, so it would be valuable to explicitly explain what the difference between those two things is

**Review by suo — COMMENTED** [review]
reviewing the alias analysis bits

File: torch/csrc/jit/passes/alias_analysis.cpp (hunk 25 above)
Comment (eellison) [reply torch/csrc/jit/passes/alias_analysis.cpp]: yes - as the comment says, this is an important api, so we shouldn't return True for a non-aliasing value if there is a wildcard write in the graph

File: torch/csrc/jit/passes/alias_analysis.h (hunk 26 above)
Comment (eellison) [reply torch/csrc/jit/passes/alias_analysis.h]: The more descriptive comment in is the cpp comment, and this is a private method.

File: torch/csrc/jit/script/exit_transforms.cpp [hunk 27]
```diff
@@ -76,11 +78,58 @@ struct ExitTransformer {
...
+    } else {
+      current_exit_kind_ = prim::LoopContinuation;
+      convertLoopOutputsToContinuations(graph_->block());
+    }
     transformExits(graph_->block());
   }
 
  private:
+  // The Logic for the exiy transform simplifies if the block outputs
+  // are converted to Exits before running, because you do not
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/exit_transforms.cpp]: This comment is a bit unclear: is it a suggestion to improve (~simplify) the implementation in future, or is it an assumption we're already making? Also, it might help if we add a brief description of what function is expected to do.

File: torch/csrc/jit/script/exit_transforms.cpp [hunk 28]
```diff
@@ -76,11 +78,58 @@ struct ExitTransformer {
...
+        convertLoopOutputsToContinuations(b);
+      }
+    }
+    if (owningNodeKind(block) == prim::Loop) {
+      convertBlockOutputsToNode(block, prim::LoopContinuation);
+    }
+  }
+
+  static bool isGraphOrClosureBlock(Block* block) {
+    return owningNodeKind(block) == prim::Function ||
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/exit_transforms.cpp]: Nitpick: even though `owningNodeKind` correctly handles blocks without the owning node, I'd swap the expressions in this `||` statement.

File: torch/csrc/jit/script/exit_transforms.cpp [hunk 29]
```diff
@@ -125,6 +161,48 @@ struct ExitTransformer {
...
+    new_if->addBlock()->registerOutput(loop.nextCond());
+    auto new_condition = new_if->addOutput()->setType(BoolType::get());
+    loop.bodyBlock()->eraseOutput(0);
+    loop.bodyBlock()->insertOutput(0, new_condition);
+
+    // we also need to propagate hasExited() and exitValues() outside the loop
+
+    // we didn't exit if we didn't enter the loop
+    node->addInput(false_val_);
+    body->addInput()->setType(BoolType::get());
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/exit_transforms.cpp]: It's pretty hard to track what's going on in this function. Could you please add a description or an example?

File: torch/csrc/jit/script/exit_transforms.cpp [hunk 30]
```diff
@@ -254,45 +332,73 @@ struct ExitTransformer {
...
 
+  // if we are transforming LoopContinuations, then when we enter a loop
+  // the exits target that loop. if transforming ReturnStmts, exit target
+  // the most recent Closure/Graph
+  void setTargetBlock(Block* block) {
+    if (current_exit_kind_ == prim::LoopContinuation) {
+      if (owningNodeKind(block) == prim::Loop) {
+        target_block_ = block;
+      }
+    } else if (current_exit_kind_ == prim::ReturnStmt) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/exit_transforms.cpp]: It is possible to have `current_exit_kind_ == prim::LoopContinuation` but `owningNodeKind(block) != prim::Loop `, right? In this case we're just not updating the `target_block_`? Is my understanding correct? If yes, could you please add this to the comment and also maybe rename this function to `update` instead of `set`.

File: torch/csrc/jit/script/exit_transforms.cpp [hunk 31]
```diff
@@ -254,45 +332,73 @@ struct ExitTransformer {
...
-      if (status == MIGHT) {
+      if (status == ExitStatus::MIGHT) {
         if (it != block->nodes().end()) {
           exit_pair = guardBlockNodes(block, exit_pair, it);
         }
         break;
       }
     }
+
+    // if we are targeting this block, update the output values to the
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/exit_transforms.cpp]: Please explain in which cases this is expected to happen.

File: torch/csrc/jit/script/exit_transforms.cpp [hunk 32]
```diff
@@ -310,41 +416,24 @@ struct ExitTransformer {
 
   // we create one uninitialized value per type, cache it here and reuse it
   std::unordered_map<TypePtr, Value*> unit_values_;
+
+  // can either be LoopContinuation/ReturnStmt
+  Symbol current_exit_kind_;
   Value* true_val_;
   Value* false_val_;
 
+  // the block that we are currently targeting in the transform.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/exit_transforms.cpp]: Might be helpful to provide some context here - it's not obvious what "targeting" means unless you already know what the transformation is doing.

File: torch/csrc/jit/script/compiler.cpp [hunk 33]
```diff
@@ -545,6 +545,21 @@ struct to_ir {
     return old_frame;
   }
 
+  // If the graph might not return, at an implicit None return at the end
```
Comment (ZolotukhinM) [inline torch/csrc/jit/script/compiler.cpp]: s/at/add/ ?

**Review by ZolotukhinM — APPROVED** [review]
I have mostly nitpicks and comments suggestions/requests and other than that the code looks good to me (I looked at this commit: https://github.com/pytorch/pytorch/pull/19179/commits/856681435dc202d83e3a36c10e41c090529574c5).

The transformation we're doing here is really convoluted, so I'd like to make sure we have all the steps are well commented plus we have a document explaining the overall design. Please follow up on this (not necessary in this PR).

File: torch/csrc/jit/script/exit_transforms.cpp (hunk 27 above)
Comment (eellison) [reply torch/csrc/jit/script/exit_transforms.cpp]: On second thought, it doesn't make it that much more simpler... I think I will remove it but i will leave that for a follow up.

**Review by eellison — COMMENTED** [review]
thanks for thorough review!
